lib/src/protobuf/json.dart

```diff
@@ -68,10 +76,26 @@ Map<String, dynamic> _writeToJsonMap(_FieldSet fs) {
 // (Called recursively on nested messages.)
 void _mergeFromJsonMap(
     _FieldSet fs, Map<String, dynamic> json, ExtensionRegistry registry) {
+  return _mergeFromJsonMapWithMapper(
+      fs, json, registry, (meta, key) => meta.byTagAsString[key]);
+}
+
+// Merge fields from a previously decoded JSON object.
+// (Called recursively on nested messages.)
+void _mergeFromProto3JsonMap(
+    _FieldSet fs, Map<String, dynamic> json, ExtensionRegistry registry) {
+  return _mergeFromJsonMapWithMapper(
+      fs, json, registry, (meta, key) => meta.byName[key]);
+}
+
+// Merge fields from a previously decoded JSON object.
+// (Called recursively on nested messages.)
+void _mergeFromJsonMapWithMapper(_FieldSet fs, Map<String, dynamic> json,
+    ExtensionRegistry registry, Function mapper) {
```

Comment:
Pass in meta.byName or meta.byTagAsString.
This avoids the allocation for the closure that is passed in.
Call the parameter keyToFieldInfo, not the generic name 'mapper'.
```suggestion
    ExtensionRegistry registry, Map<String, FieldInfo> keyToFieldInfo) {
```

I'd also rename `_mergeFromJsonMapWithMapper` to something like `_mergeFromJsonMapWithFieldMap`.